File: rust_c/temporal-safety-unsafe/box_c.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    // Allocate an integer on the heap using malloc
    int *boxed_data = (int *)malloc(sizeof(int));
    *boxed_data = 42;

    // Get a raw pointer to the data inside the Box
    int *raw_ptr = boxed_data;

    // Simulate freeing the memory (use-after-free error)
    free(raw_ptr);

    // Attempt to access the freed memory using the raw pointer (unsafe)
    // This is a use-after-free error and can lead to undefined behavior
    //printf("Data after dropping: %d\n", *raw_ptr);

    return *raw_ptr;
}

File: rust_c/temporal-safety-unsafe/box_c.rs
fn main() {
    let mut boxed_data = Box::new(42);
    let raw_ptr = Box::into_raw(boxed_data);

    // freeing the memory (use-after-free error)
    unsafe { Box::from_raw(raw_ptr); }

    // Attempt to access the freed memory using the raw pointer (unsafe)
    // This is a use-after-free error and can lead to undefined behavior
    unsafe {
        println!("Data after dropping: {}", *raw_ptr);
    }
}